lib/util/search.js

```diff
@@ -130,31 +134,94 @@ SearchClient.prototype.parseUri = function(uri) {
   }
 }
 
-SearchClient.prototype.buildIndex = async function() {
+SearchClient.prototype.createIndexName = function() {
+  const datetime = moment().format('YYYYMMDDHHmmss')
+  return `${this.indexNames.base}-${datetime}`
+}
+
+SearchClient.prototype.createIndex = async function(index) {
   await this.checkESVersion()
   const mappingDir = this.crowi.resourceDir + 'search'
   const mappingFile = path.join(mappingDir, this.selectMappingFile())
   const body = require(mappingFile)
-  return this.client.indices.create({
-    index: this.indexName,
-    body,
-  })
+  return this.client.indices.create({ index, body })
 }
 
-SearchClient.prototype.deleteIndex = function(uri) {
-  return this.client.indices.delete({
-    index: this.indexName,
-  })
+SearchClient.prototype.buildIndex = async function(newIndexName) {
+  await this.addAllPages(newIndexName)
+
+  const alias = await this.getAlias()
+  const add = {
+    index: newIndexName,
+    alias: this.indexNames.current,
+  }
+  if (alias) {
+    const remove = {
+      index: alias.index,
+      alias: alias.alias,
+    }
+    await this.updateAliases({ add, remove })
+  } else {
+    await this.updateAliases({ add })
+  }
+
+  const indices = await this.getIndices()
+  const deleteIndices = indices.filter(index => index !== newIndexName)
+  await this.deleteIndices(deleteIndices)
+}
+
+SearchClient.prototype.getIndices = async function() {
+  const indices = await this.client.cat.indices({ format: 'json' })
+  return indices.map(({ index }) => index).filter(index => index.startsWith(this.indexNames.base))
+}
+
+SearchClient.prototype.deleteIndices = function(indices) {
+  return this.client.indices.delete({ index: indices })
+}
+
+SearchClient.prototype.existsAlias = function() {
+  return this.client.indices.existsAlias({ name: this.indexNames.current })
+}
+
+SearchClient.prototype.getAlias = function() {
+  const existsAlias = this.existsAlias()
+  if (existsAlias) {
+    const aliases = this.client.cat.aliases({ name: this.indexNames.current, format: 'json' })
+    if (aliases.length > 0) {
+      return aliases[0]
+    }
+  }
+  return null
+}
+
+SearchClient.prototype.putAlias = async function(index) {
+  return this.client.indices.putAlias({ index, name: this.indexNames.current })
+}
+
+SearchClient.prototype.ensureAlias = async function() {
+  const exists = await this.existsAlias()
+  if (!exists) {
+    const indices = await this.getIndices()
+    if (indices.length > 0) {
```

Comment:
📝memo: If search index is never built yet, this gonna false and this function returns `false`.